src/res/mod.rs

```diff
@@ -244,124 +257,295 @@ impl Resources {
-    pub fn fetch<T>(&self) -> Fetch<T>
+    pub fn fetch<T, F, P>(&self, with: F) -> Fetch<T, P>
     where
         T: Resource,
+        F: FnOnce() -> T,
     {
-        self.try_fetch().expect(RESOURCE_NOT_FOUND)
+        let res_id = ResourceId::new::<T>();
+        let r = self.resources
+            .get(&res_id)
+            .unwrap_or_else(|| self.def_res(res_id, with()));
+
+        Fetch {
+            inner: r.borrow(),
+            phantom: PhantomData,
+        }
     }
 
-    /// Like `fetch`, but returns an `Option` instead of panicking in the case of the resource
-    /// being accessed mutably.
-    pub fn try_fetch<T>(&self) -> Option<Fetch<T>>
+    /// Like `fetch`, but returns an `Option` instead of inserting a default value
+    /// in case the resource does not exist.
+    pub fn try_fetch<T, P>(&self) -> Option<Fetch<T, P>>
     where
         T: Resource,
     {
-        self.try_fetch_internal(TypeId::of::<T>()).map(|r| {
-            Fetch {
+        let res_id = ResourceId::new::<T>();
+
+        self.resources
+            .get(&res_id)
+            .or_else(|| self.get_res::<T>(&res_id))
+            .map(|r| Fetch {
                 inner: r.borrow(),
                 phantom: PhantomData,
-            }
-        })
+            })
     }
 
     /// Fetches the resource with the specified type `T` mutably.
     ///
     /// Please see `fetch` for details.
-    pub fn fetch_mut<T>(&self) -> FetchMut<T>
+    pub fn fetch_mut<T, F, P>(&self, with: F) -> FetchMut<T, P>
     where
         T: Resource,
+        F: FnOnce() -> T,
     {
-        self.try_fetch_mut().expect(RESOURCE_NOT_FOUND)
+        let res_id = ResourceId::new::<T>();
+        let r = self.resources
+            .get(&res_id)
+            .unwrap_or_else(|| self.def_res::<T>(res_id, with()));
+
+        FetchMut {
+            inner: r.borrow_mut(),
+            phantom: PhantomData,
+        }
     }
 
-    /// Like `fetch_mut`, but returns an `Option` instead of panicking in the case of the resource
-    /// being accessed mutably.
-    pub fn try_fetch_mut<T>(&self) -> Option<FetchMut<T>>
+    /// Like `fetch_mut`, but returns an `Option` instead of inserting a default value
+    /// in case the resource does not exist.
+    pub fn try_fetch_mut<T, P>(&self) -> Option<FetchMut<T, P>>
     where
         T: Resource,
     {
-        self.try_fetch_internal(TypeId::of::<T>()).map(|r| {
-            FetchMut {
+        let res_id = ResourceId::new::<T>();
+
+        self.resources
+            .get(&res_id)
+            .or_else(|| self.get_res::<T>(&res_id))
+            .map(|r| FetchMut {
                 inner: r.borrow_mut(),
                 phantom: PhantomData,
-            }
-        })
+            })
     }
 
-    fn try_fetch_internal(&self, id: TypeId) -> Option<&TrustCell<Box<Resource>>> {
-        self.resources.get(&ResourceId(id))
+    fn def_res<T>(&self, res_id: ResourceId, value: T) -> &TrustCell<Box<Resource>>
```

Comment:
this is probably the main problem I have with the PR - introducing another state of resources "added but not yet maintained" with an associated deferred resource storage